File: 06-breadth-first-search/python/breadth_first_search.py
from collections import deque

def breadth_first_search(graph, start, finish):
    """Implementation of breadth-first search algorithm"""
    search_queue = deque(graph[start])
    searched = {}

    while search_queue:
        node = search_queue.popleft()
        if not searched.get(node):
            if node == finish:
                return (True, node)
            else:
                search_queue += graph[node]
                searched[node] = True
    return (False, None)

File: 06-breadth-first-search/ruby/breadth_first_search.rb
# frozen_string_literal: true

def breadth_first_search(graph, start, finish)
  search_queue = graph[start].dup
  searched = {}

  loop do
    break if search_queue.empty?
    node = search_queue.shift
    next if searched[node]

    return [true, node] if node == finish
    search_queue += graph[node]
    searched[node] = true
  end
  return [false, nil]
end
